crates/cheatcodes/src/inspector.rs

```diff
@@ -590,23 +590,29 @@ impl<DB: DatabaseExt + Send> Inspector<DB> for Cheatcodes {
     }
 
     fn step_end(&mut self, interpreter: &mut Interpreter<'_>, data: &mut EVMData<'_, DB>) {
-        if self.use_zk_vm && interpreter.current_opcode() == opcode::BALANCE {
-            if interpreter.stack.is_empty() {
-                interpreter.instruction_result = InstructionResult::StackUnderflow;
-                return;
-            }
-            // Safety: Length is checked above.
-            let address = Address::from_word(B256::from(unsafe { interpreter.stack.pop_unsafe() }));
-            let balance = foundry_zksync_core::balance(address, data.db, &mut data.journaled_state);
-
-            // Skip the current BALANCE instruction since we've already handled it
-            match interpreter.stack.push(balance) {
-                Ok(_) => unsafe {
-                    interpreter.instruction_pointer = interpreter.instruction_pointer.add(1);
-                },
-                Err(e) => {
-                    interpreter.instruction_result = e;
+        let address = match interpreter.current_opcode() {
+            opcode::SELFBALANCE if self.use_zk_vm => interpreter.contract().address,
```

Comment:
Makes sense to move `if self.use_zk_vm` to top level as this entire block only applies in that scenario and makes it easier to add other opcodes in future